File: 0235.Lowest-Common-Ancestor-of-a-Binary-Search-Tree/step2.cpp
// Definition for a binary tree node.
#include <algorithm>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
  public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        int smaller = std::min(p->val, q->val);
        int larger = std::max(p->val, q->val);
        TreeNode *node = root;
        while (node) {
            if (smaller <= node->val && node->val <= larger) {
                return node;
            }
            if (node->val < smaller) {
                node = node->right;
                continue;
            }
            if (node->val > larger) {
                node = node->left;
                continue;
            }
        }
        return nullptr;
    }
};

File: 0235.Lowest-Common-Ancestor-of-a-Binary-Search-Tree/step2.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        smaller, larger = sorted((p.val, q.val))
        node = root
        while True:
            if smaller <= node.val <= larger:
                return node
            if node.val < smaller:
                node = node.right
                continue
            if node.val > larger:
                node = node.left
        return None
